src/ui/draggable/draggable.ts

```diff
@@ -0,0 +1,150 @@
+import { DraggableSynchronizer } from './draggable-synchronizer';
+import { dom, DomWatcher, Raf } from '../..';
+import { Vector } from '../../mathf/vector';
+import { CachedMouseTracker } from '../../dom/cached-mouse-tracker';
+
+/**
+ * Type of a function that can be used to constrain the movement of a vector.
+ */
+export type DraggableConstraint =
+    (draggable: Draggable, delta: Vector) => Vector;
+
+/**
+ * Events to dispatch on the start and end of a drag.
+ */
+export enum DraggableEvent {
+  START = 'deguDraggableStart',
+  END = 'deguDraggableEnd'
+}
+
+/**
+ * Makes a DOM element draggable by adjusting the element's transform.
+ */
+export class Draggable {
+  readonly element: HTMLElement;
+  protected mouseTracker: CachedMouseTracker;
+  private constraints: DraggableConstraint[];
+  private readonly raf: Raf;
+  private readonly draggableSynchronizer: DraggableSynchronizer;
+  private lastPosition: Vector;
+  private domWatcher: DomWatcher;
+
+  constructor(
+    element: HTMLElement,
+    { constraints = [] }: { constraints?: DraggableConstraint[] } = {}
+  ) {
+    this.element = element;
+    this.raf = new Raf(() => this.loop());
+    this.lastPosition = null;
+    this.constraints = [...constraints];
+    this.mouseTracker = CachedMouseTracker.getSingleton(this);
+    this.draggableSynchronizer = DraggableSynchronizer.getSingleton(this);
+    this.domWatcher = new DomWatcher();
+    this.init();
+  }
+
+  /**
+   * Disposes of the Draggable instance.
+   */
+  dispose() {
+    this.raf.stop();
+    this.constraints = [];
+    this.mouseTracker.dispose(this);
+    this.draggableSynchronizer.dispose(this);
```

Comment:
See my comment on Carousel and how I think we should avoid this pattern of multiple singletons.

Reply:
Responded in the carousel comment.